tests/flow/test_hnswlib.py

```diff
@@ -518,11 +450,481 @@ def test_multi_range_query():
+
+    start = time.time()
+    res_labels_parallel, res_dists_parallel = parallel_multi_index.knn_parallel(query_data, k, num_threads=n_threads)
+    total_search_time_parallel = time.time() - start
+
+    total_correct_parallel = 0
+    for res_labels, ground_truth in zip(res_labels_parallel, total_res_bf):
+        total_correct_parallel += len(set(res_labels).intersection(set(ground_truth)))
+
+    print(f"Running parallel search, got {total_correct_parallel / (k * num_queries)} recall on {num_queries} queries,"
+          f" average query time is {total_search_time_parallel / num_queries} seconds")
+    print(f"Got {total_search_time / total_search_time_parallel} times improvement in runtime using"
+          f" {n_threads} threads\n")
+    assert total_correct_parallel >= total_correct*0.95
+    assert total_search_time/total_search_time_parallel >= expected_speedup
+
+    # Insert vectors to the index and search in parallel.
+    parallel_multi_index = create_hnsw_index(dim, num_elements, metric, data_type, m=32, ef_runtime=200,
+                                             is_multi=True)
+    assert parallel_multi_index.index_size() == 0
+
+    def insert_vectors():
+        parallel_multi_index.add_vector_parallel(data, labels, num_threads=int(n_threads/2))
+
+    res_labels_g = np.zeros((num_queries, dim))
+
+    def run_queries():
+        nonlocal res_labels_g
+        res_labels_g, _ = parallel_multi_index.knn_parallel(query_data, k, num_threads=int(n_threads/2))
+
+    t_insert = threading.Thread(target=insert_vectors)
+    t_query = threading.Thread(target=run_queries)
+    print("Running queries in parallel to inserting vectors to the index, start running queries after more 50% of the"
+          " vectors are indexed")
+    t_insert.start()
+    # Wait until half of the index is indexed, then start run queries
+    while parallel_multi_index.index_size() < num_elements / 2:
+        time.sleep(0.5)
+    t_query.start()
+
+    [t.join() for t in [t_insert, t_query]]
+
+    # Measure recall - expect to get increased recall over time, since vectors are being inserted while queries
+    # are running, and the ground truth is measured compared to the index that contains all the elements.
+    chunk_size = int(num_queries/5)
+    total_correct_prev_chunk = 0
+    for i in range(0, num_queries, chunk_size):
+        total_correct_cur_chunk = 0
+        for j in range(i, i+chunk_size):
+            total_correct_cur_chunk += len(set(total_res_bf[j]).intersection(set(res_labels_g[j])))
+        if i == chunk_size:  # first iteration, there is no previous chunk
+            total_correct_prev_chunk = total_correct_cur_chunk
+        else:
+            assert total_correct_cur_chunk >= total_correct_prev_chunk
+            total_correct_prev_chunk = total_correct_cur_chunk
+        print(f"Recall for chunk {int(i/chunk_size)+1}/{int(num_queries/chunk_size)} of queries is:"
```

Comment:
```suggestion
        print(f"Recall for queries' chunk {int(i/chunk_size)+1}/{int(num_queries/chunk_size)} is:"
```
Took me a while to understand this sentence 😅